### PR TITLE
Minesweeper

### PR DESCRIPTION
**Pull request by pokrakam**
@larshp as discussed some funky stuff going on with the test runner on this one.

### CODE REVIEW

File: src/zcl_minesweeper.abap [hunk 1]
```diff
@@ -0,0 +1,75 @@
+class zcl_minesweeper definition
+  public
+  final
```
Comment (larshp) [inline src/zcl_minesweeper.abap]: try renaming this file, from "zcl_minesweeper.abap" to "zcl_minesweeper.clas.abap"

Comment (pokrakam) [reply src/zcl_minesweeper.abap]: D'Oh!
Thanks, updating PR with next bug (a real one this time!)

**Comment by pokrakam** [comment]
OK, with correct filename there's an issue in downport. How far down is it ported? `reduce` seems to work but a `let` variable is pulled out of the `reduce` statement into the method body.
```abap
    result = REDUCE i(
        INIT s = 0
        FOR i = nmax( val1 = 1 val2 = y - 1 ) UNTIL
            i > nmin( val1 = y + 1 val2 = height )
        LET row = board[ i ] IN                    "<==== Downport extracts this out of reduce
        NEXT s = s + count( val = row+off(len)
                            sub = `*` ) ).
```

downported to:
```abap
    DATA temp5 LIKE LINE OF board.
...
    READ TABLE board INDEX i INTO temp5.      "<==== oops!
    IF sy-subrc <> 0.
      RAISE EXCEPTION TYPE cx_sy_itab_line_not_found.
    ENDIF.
    result = REDUCE i(
        INIT s = 0
        FOR i = nmax( val1 = 1 val2 = y - 1 ) UNTIL
            i > nmin( val1 = y + 1 val2 = height )
        LET row = temp5 IN 
        NEXT s = s + count( val = row+off(len)
                            sub = `*` ) ).
```

**Comment by larshp** [comment]
need abaplint to also downport REDUCE

**Comment by pokrakam** [comment]
Unclear what you mean: Change solution to work without, or will this come to exercism?

**Comment by larshp** [comment]
https://github.com/abaplint/abaplint/issues/2336 opened

**Comment by pokrakam** [comment]
Also downport is not happy about implicit type conversions. 
`string_var = cond #( if some_int = 0 then space else some_int ).`
(already adjusted code in this example, see last commented line in zcl_minesweeper for the failing version) 
Downport generates a temp var of type undefined. 
Is this a test bug in this repo or abaplint issue?

**Comment by larshp** [comment]
in abaplint repo

**Comment by pokrakam** [comment]
Do you still need this PR? The downport still fails.

**Comment by larshp** [comment]
havent had the time to look into it...

**Comment by pokrakam** [comment]
OK no prob, no rush my end. Wasn't sure if that's what your comment meant, and have the other test issue.

**Comment by larshp** [comment]
first bug to fix in the linter

```abap
DATA input TYPE STANDARD TABLE OF string WITH DEFAULT KEY.
input  = VALUE #( ( ) ).
```

https://github.com/abaplint/abaplint/pull/2460

**Comment by larshp** [comment]
next, bug in transpiler,

```abap
DATA lv TYPE i.
lv = count( val = 'ab*foo*cd' sub = `*` ).
ASSERT lv = 2.
```

https://github.com/abaplint/transpiler/pull/661

**Comment by larshp** [comment]
next transpiler bug,

```abap
DATA row TYPE c LENGTH 1.
DATA result TYPE string.
result = |{ row WIDTH = 10 }|.
WRITE strlen( result ).
```

https://github.com/abaplint/transpiler/pull/662

**Comment by larshp** [comment]
next transpiler bug, 😁 

```abap
DATA row TYPE c LENGTH 10.
row = ' * '.
row(1) = '1'.
ASSERT row+2(1) = space.
```

https://github.com/abaplint/transpiler/pull/663

**Comment by larshp** [comment]
running 🎉 

image

**Pull request merged**